server/asset/eth/rpcclient.go

```diff
@@ -124,9 +129,55 @@ func (c *rpcclient) transaction(ctx context.Context, hash common.Hash) (tx *type
 // accountBalance gets the account balance, including the effects of known
 // unmined transactions.
 func (c *rpcclient) accountBalance(ctx context.Context, addr common.Address) (*big.Int, error) {
-	bigBal, err := c.ec.PendingBalanceAt(ctx, addr)
+	tip, err := c.blockNumber(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("blockNumber error: %v", err)
+	}
+	currentBal, err := c.ec.BalanceAt(ctx, addr, big.NewInt(int64(tip)))
 	if err != nil {
 		return nil, err
 	}
-	return bigBal, nil
+
+	var txs map[string]map[string]*RPCTransaction
+	err = c.c.CallContext(ctx, &txs, "txpool_contentFrom", addr)
+	if err != nil {
+		return nil, fmt.Errorf("contentFrom error: %w", err)
+	}
+
+	outgoing := new(big.Int)
+	for _, group := range txs { // 2 groups, pending and queued
+		for _, tx := range group {
+			outgoing.Add(outgoing, tx.Value.ToInt())
+			gas := new(big.Int).SetUint64(uint64(tx.Gas))
+			if tx.GasPrice != nil && tx.GasPrice.ToInt().Cmp(bigZero) > 0 {
+				outgoing.Add(outgoing, new(big.Int).Mul(gas, tx.GasPrice.ToInt()))
+			} else if tx.GasFeeCap != nil {
+				outgoing.Add(outgoing, new(big.Int).Mul(gas, tx.GasFeeCap.ToInt()))
+			}
+		}
+	}
+
+	return currentBal.Sub(currentBal, outgoing), nil
```

Comment:
"smarter" in the sense that it excludes pending incoming?

Reply:
I guess. I think we have to, no?

- Lot size 1 eth
- User has current balance 1.5 eth
- Pending balance also 1.5, but with 1 incoming transfer of 1 eth and one outgoing transfer of 1 eth. 
- Place order.
- The outgoing gets mined but not the incoming. current balance = 0.5, pending = 1.5. 
- Match as maker.
- Can't swap.

Reply:
My bad. I didn't mean to drop this out of nowhere. I thought this was a revelation from the Matrix convo last week, but maybe I misunderstood. Is my read right, though?

Reply:
Sure sure, I'm not totally surprised to see the change given what we've been talking about, but there's no code comments about the intent of all this code.

One might naively just use `PendingBalanceAt` or `BalanceAt` of course so I think there needs to be explanation.  When I see `"txpool_contentFrom"` and a us cloning some json-tagged type `RPCTransaction`, I raise an eyebrow.  It's surprising to me that we're the first ones to need to do this kind of accounting or that there are no straightforward methods and types ready for this.

I say fine, lets run with it, but it deserves some docs and serious consideration because it just smells very low level.